File: development/libraries/falcon_packet/ff_encoder.c
#include "ff_encoder.h"
#include "fp_decode.h"
#include "fp_encode.h"
#include "checksum.h"

#include <stdlib.h>
#include <string.h>

#define FRAME_SIZE_BYTE_INDEX (FRAME_HEADER_SIZE-1)

void ff_encoder_init(ff_encoder_t *encoder)
{
  encoder->destBuffer = NULL;
  encoder->writeIndex = 0;
}

void ff_encoder_set_buffer(ff_encoder_t *encoder, uint8_t *buffer)
{
  encoder->destBuffer = buffer;
  buffer[0] = FRAME_DELIMITER1;
  buffer[1] = FRAME_DELIMITER2;
  buffer[2] = 0;
  encoder->writeIndex = FRAME_HEADER_SIZE;
}

int ff_encoder_append_packet(ff_encoder_t *encoder, void *packet, fp_type_t packetType)
{
  uint8_t packetLength = fp_get_packet_length(packetType) + PACKET_HEADER_SIZE;

  if ((encoder->writeIndex + packetLength) > (MAX_FRAME_SIZE - FRAME_FOOTER_SIZE)) {
    return -1;
  }

  uint8_t bytesWritten = fp_encode_packet(encoder->destBuffer + encoder->writeIndex, packet, packetType);
  if (bytesWritten == packetLength) {
    encoder->writeIndex += bytesWritten;
    return bytesWritten;
  }
  return -1;
}

int ff_encoder_append_data(ff_encoder_t *encoder, uint8_t *data, uint32_t length)
{
  if (length > MAX_PACKET_SIZE) {
    return -1;
  }
  if ((encoder->writeIndex + length) > (MAX_FRAME_SIZE - FRAME_FOOTER_SIZE)) {
    return -1;
  }

  memcpy(encoder->destBuffer + encoder->writeIndex, data, length);
  encoder->writeIndex += length;
  return length;
}

int ff_encoder_append_footer(ff_encoder_t *encoder)
{
  uint16_t crc = crc_modbus(&encoder->destBuffer[FRAME_HEADER_SIZE], encoder->writeIndex - FRAME_HEADER_SIZE);
  encoder->destBuffer[encoder->writeIndex] = (crc >> 0) & 0xFF;
  encoder->writeIndex++;
  encoder->destBuffer[encoder->writeIndex] = (crc >> 8) & 0xFF;
  encoder->writeIndex++;

  /* Write size byte */
  encoder->destBuffer[FRAME_SIZE_BYTE_INDEX] = encoder->writeIndex;
  return encoder->writeIndex;
}

uint32_t ff_encoder_get_remaining_bytes(ff_encoder_t *encoder)
{
  return MAX_FRAME_SIZE - encoder->writeIndex - FRAME_FOOTER_SIZE;
}

File: development/libraries/falcon_packet/ff_encoder.py
import struct
import libscrc
from falcon_packet import *

class FF_Encoder():
    def __init__(self):
        pass

    def write_frame_header(self, dest, size):
        struct.pack_into('<B', dest, 0, FRAME_DELIMITER1)
        struct.pack_into('<B', dest, 1, FRAME_DELIMITER2)
        struct.pack_into('<B', dest, 2, size)
        return FRAME_HEADER_SIZE

    def write_frame_footer(self, dest, offset, crc16):
        struct.pack_into('<B', dest, offset, crc16 & 0xFF)
        struct.pack_into('<B', dest, offset + 1, (crc16 >> 8) & 0xFF)
        return FRAME_FOOTER_SIZE

    def append_frame_packet(self, dest, offset, packet):
        encodedPacket = packet.encode()
        packetLength = len(encodedPacket)
        assert((offset + packetLength) < (MAX_FRAME_SIZE - FRAME_FOOTER_SIZE))
        dest[offset:offset+packetLength] = encodedPacket
        return packetLength

    def pack_packets_into_frame(self, packets):
        retbuf = bytearray(MAX_FRAME_SIZE)

        writeIndex = FRAME_HEADER_SIZE
        for packet in packets:
            writeIndex += self.append_frame_packet(retbuf, writeIndex, packet)


        crc16 = libscrc.modbus(retbuf[FRAME_HEADER_SIZE:writeIndex])
        writeIndex += self.write_frame_footer(retbuf, writeIndex, crc16)

        self.write_frame_header(retbuf, writeIndex)
        return retbuf

